### PR TITLE
ARROW-9892: [Rust] [DataFusion] Added concat of utf8

### PR DESCRIPTION
**Pull request by jorgecarleitao**

### CODE REVIEW

File: rust/datafusion/src/execution/physical_plan/string_expressions.rs [hunk 1]
```diff
@@ -0,0 +1,63 @@
+// Licensed to the Apache Software Foundation (ASF) under one
```
Comment (jorgecarleitao) [inline rust/datafusion/src/execution/physical_plan/string_expressions.rs:1]: I created a new one to not over-populate `expressions.rs`. If we accept this, we should prob. move `length` here.

File: rust/datafusion/src/execution/physical_plan/string_expressions.rs [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+            })
+    }};
+}
+
+/// concatenate string columns together.
+pub fn concatenate(args: &[ArrayRef]) -> Result<StringArray> {
+    // downcast all arguments to strings
+    let args = downcast_vec!(args, StringArray).collect::<Result<Vec<&StringArray>>>()?;
+    // do not accept 0 arguments.
+    assert!(args.len() != 0);
```
Comment (jorgecarleitao) [inline rust/datafusion/src/execution/physical_plan/string_expressions.rs:39]: this currently panics on zero arguments, which means that the plan `SELECT c1, concat() FROM t` crashes on run-time.

We should prob. safe-guard this in type coercion.

Essentially, all our physical expressions currently require one argument or they do not know how many rows to return.

Comment (alamb) [reply rust/datafusion/src/execution/physical_plan/string_expressions.rs:39]: I would prefer that we not commit code that can `panic`s -- so in this PR I would suggest doing one of the following
1. Add an error in physical planing for `concat()`
2. Implement `concat()`

Here is one possible behavior for `concat()`:
1. The result value is always the Null
2. It should produce one output value for each input value

A potential implementation could be "rewrite `concat()` to be `concat("")` -- aka have the planner insert a single constant argument

File: rust/datafusion/tests/sql.rs [hunk 3]
```diff
@@ -682,6 +682,32 @@ fn query_length() -> Result<()> {
...
+        ],
+    )?;
+
+    let table = MemTable::new(schema, vec![vec![data]])?;
+
+    let mut ctx = ExecutionContext::new();
+    ctx.register_table("test", Box::new(table));
+    let sql = "SELECT concat(c1, '-hi-', cast(c2 as varchar)) FROM test";
+    let actual = execute(&mut ctx, sql);
+    let expected = vec!["\"-hi-0\"", "\"a-hi-1\"", "\"\"", "\"aaa-hi-3\""];
```
Comment (alamb) [inline rust/datafusion/tests/sql.rs:706]: I am surprised that the third result is `""` and not `NULL` (as the third i32 input is NULL). Is that some artifact of using `execute`?

**Review by alamb — APPROVED** [review]
Looks good to me other than the panic! behavior -- I think that should be fixed before this is merged

File: rust/datafusion/src/execution/physical_plan/string_expressions.rs (hunk 2 above)
Comment (jorgecarleitao) [reply rust/datafusion/src/execution/physical_plan/string_expressions.rs:39]: Thanks for the suggestions 👍 I've added an error when the plan has 0 arguments.

File: rust/datafusion/tests/sql.rs (hunk 3 above)
Comment (jorgecarleitao) [reply rust/datafusion/tests/sql.rs]: Yeah, it took me some time to figure that out also; it is an artifact from `execute`. I changed its behavior to output NULL, but I think we may need to reformulate that code at some point, as only `utf8` show `NULL`.

**Review by andygrove — APPROVED** [review]
LGTM

**Comment by andygrove** [comment]
@jorgecarleitao This one needs rebasing now

**Pull request closed without merging**